types.go

```diff
@@ -15,33 +15,34 @@ import (
 
 const defaultGradientAbsTol = 1e-6
 
-// EvaluationType is used by the optimizer to specify information needed
-// from the objective function.
+// EvaluationType is used by the optimizer to specify information needed from
+// the objective function. The types can be composed together using binary or
+// operator.
 type EvaluationType int
 
 const (
-	NoEvaluation EvaluationType = iota
-	FuncEvaluation
-	GradEvaluation
-	HessEvaluation
-	FuncGradEvaluation
-	FuncGradHessEvaluation
+	NoEvaluation           EvaluationType = 0
```

Comment:
I'm bad with bit math (hopefully this PR will help), so pardon my ignorance.

I would like it to be that FuncGradEvaluation exists implicitly, and instead Method returns 
````
FuncEvaluation + GradEvaluation
````
to express that it wants both. This is very friendly from an implementation perspective. 

````
FuncEvaluation + GradEvaluation + ConstraintEvaluation
````
We have 4 bits (Func, grad, hessian, constraint), and so we can test that result by looking at the bit (something involving bit shift?).